s3_to_redshift/s3_to_redshift.py

```diff
@@ -329,56 +331,80 @@ def report(data):
+            try:
+                csv_string = csv_string.decode('utf-8')
+            except UnicodeDecodeError as _e:
+                report_stats['failed'] += 1
+                report_stats['bad'] += 1
+                report_stats['bad_list'].append(object_summary)
+                report_stats['incomplete_list'].remove(object_summary)
+                e_object = _e.object.splitlines()
+                logger.exception(
+                    ''.join((
+                        "Decoding UTF-8 failed for file {0}\n"
+                        .format(object_summary.key),
+                        "The input file stopped parsing after line {0}:\n{1}\n"
+                        .format(len(e_object), e_object[-1]),
+                        "Keying to badfile and stopping.\n")))
+                try:
+                    client.copy_object(
+                        Bucket="sp-ca-bc-gov-131565110619-12-microservices",
+                        CopySource=(
+                            "sp-ca-bc-gov-131565110619-12-microservices/"
+                            f"{object_summary.key}"
+                        ),
+                        Key=badfile)
+                except Exception as _e:
+                    logger.exception("S3 transfer failed. %s", str(_e))
+                report(report_stats)
+                clean_exit(1,f'Bad file {object_summary.key} in objects to process, '
+                           'no further processing.')
 
     # Check for an empty file. If it's empty, accept it as bad and skip
     # to the next object to process
     try:
         if no_header:
-            df = pd.read_csv(
-                StringIO(csv_string),
-                sep=delim,
-                index_col=False,
-                dtype=dtype_dic,
-                usecols=range(column_count),
-                header=None)
+            if csv:
+                df = pd.read_csv(
+                    StringIO(csv_string),
+                    sep=delim,
+                    index_col=False,
+                    dtype=dtype_dic,
+                    usecols=range(column_count),
+                    header=None)
+            else:
+                df = xlsx_decoded
+                    #pd.read_excel(BytesIO(body.read()))
+                    #engine="pyxlsb",
+                    #index_col=False,
+                    #dtype=dtype_dic,
+                    #usecols=range(column_count),
+                    #header=None)
         else:
-            df = pd.read_csv(
-                StringIO(csv_string),
-                sep=delim,
-                index_col=False,
-                dtype=dtype_dic,
-                usecols=range(column_count))
+            if csv:
+                df = pd.read_csv(
+                    StringIO(csv_string),
+                    sep=delim,
+                    index_col=False,
+                    dtype=dtype_dic,
+                    usecols=range(column_count))
+            else:
```

Comment:
@lclarko : do we want to keep these lines in if they are being commented out?

Reply:
I'll remove